packages/react-core/src/hooks/useGetUrl.ts

```diff
@@ -0,0 +1,54 @@
+import * as React from 'react';
+
+import { getUrl, GetUrlInput } from 'aws-amplify/storage';
+
+import { isFunction } from '@aws-amplify/ui';
+
+export type UseGetUrlInput = GetUrlInput & {
+  onError?: (error: Error) => void;
+};
+interface UseGetUrlOutput {
+  isLoading: boolean;
+  url: URL | undefined;
+  expiresAt: Date | undefined;
+}
+
+const INIT_STATE: UseGetUrlOutput = {
+  url: undefined,
+  expiresAt: undefined,
+  isLoading: true,
+};
+
+export const useGetUrl = (input: UseGetUrlInput): UseGetUrlOutput => {
+  const [result, setResult] = React.useState(() => INIT_STATE);
+  React.useEffect(() => {
+    const { onError, ...getUrlInput } = input;
+    let ignore = false;
+
+    getUrl(getUrlInput)
+      .then((response) => {
+        if (ignore) {
+          return;
+        }
+
+        setResult({ ...response, isLoading: false });
+      })
+      .catch((error: Error) => {
+        if (ignore) {
+          return;
+        }
+
+        if (isFunction(onError)) {
+          onError(error);
+        }
+
+        setResult({ ...INIT_STATE, isLoading: false });
+      });
+
+    return () => {
```

Comment:
This won't work as expected I think because `let ignore = false;` defines a brand new one on each call

Reply:
`ignore` is set to `true` in the cleanup so if the component unmounts before the promise is resolved, the result won't be set. Then when the component re-renders we expect it to be set back to `false` so we can continue with setting the state and the most recent result will be used.

Reply:
I see. That makes sense